crypto/chacha/asm/chacha-armv8-sve.pl

```diff
@@ -0,0 +1,1008 @@
+
+sub SVE_TRANSFORMS() {
+	&add_states();
+	&transpose($xa0,$xb0,$xc0,$xd0);
+	&transpose($xa1,$xb1,$xc1,$xd1);
+	&transpose($xa2,$xb2,$xc2,$xd2);
+	&transpose($xa3,$xb3,$xc3,$xd3);
+$code.=<<___;
+	cmp	$blocks,$veclen
+	b.ge	10f
+	lsl	$dlen,$blocks,4
+	whilelo	p1.s,xzr,$dlen
+	cntp	$rsize,p1,p1.s
+___
+	&transpose($xa0,$xa1,$xa2,$xa3);
+	&transpose($xb0,$xb1,$xb2,$xb3);
+	&transpose($xc0,$xc1,$xc2,$xc3);
+	&transpose($xd0,$xd1,$xd2,$xd3);
+	&output_incr($xa0,$xa1,$xa2,$xa3);
+	&output_incr($xb0,$xb1,$xb2,$xb3);
+	&output_incr($xc0,$xc1,$xc2,$xc3);
+	&output_incr($xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	add	$ctr,$ctr,$blocks
+	b	100f
+10:
+	ptrues p1.s,ALL
+___
+	&load_8x($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7);
+	&transpose($xa0,$xa1,$xa2,$xa3);
+	&transpose($xb0,$xb1,$xb2,$xb3);
+$code.=<<___;
+	eor	$xa0.d,$xa0.d,$xt0.d
+	eor	$xa1.d,$xa1.d,$xt1.d
+	eor	$xa2.d,$xa2.d,$xt2.d
+	eor	$xa3.d,$xa3.d,$xt3.d
+	eor	$xb0.d,$xb0.d,$xt4.d
+	eor	$xb1.d,$xb1.d,$xt5.d
+	eor	$xb2.d,$xb2.d,$xt6.d
+	eor	$xb3.d,$xb3.d,$xt7.d
+___
+	&transpose($xc0,$xc1,$xc2,$xc3);
+	&store_8x($xa0,$xa1,$xa2,$xa3,$xb0,$xb1,$xb2,$xb3);
+	&load_8x($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7);
+	&transpose($xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	eor	$xc0.d,$xc0.d,$xt0.d
+	eor	$xc1.d,$xc1.d,$xt1.d
+	eor	$xc2.d,$xc2.d,$xt2.d
+	eor	$xc3.d,$xc3.d,$xt3.d
+	eor	$xd0.d,$xd0.d,$xt4.d
+	eor	$xd1.d,$xd1.d,$xt5.d
+	eor	$xd2.d,$xd2.d,$xt6.d
+	eor	$xd3.d,$xd3.d,$xt7.d
+___
+	&store_8x($xc0,$xc1,$xc2,$xc3,$xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	add	$ctr,$ctr,$veclen
+100:
+___
+}
+}}}
+
+sub SVE_LOAD_STATES() {
+	my ($tmpw0,$tmpw1,$tmpw2,$tmpw3) = ("w9","w10","w11","w12");
+
+$code.=<<___;
+	index	@mx[12].s,$ictr,1
+	// FIXME following code are not functionally necessary
+	// but appear to be vital to perf test
+#if 1
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+#endif
+___
+$code.=<<___;
+	ldp	$tmpw0,$tmpw1,[$state]
+	ldp	$tmpw2,$tmpw3,[$state,#8]
+	dup	@mx[0].s,$tmpw0
+	dup	@mx[1].s,$tmpw1
+	dup	@mx[2].s,$tmpw2
+	dup	@mx[3].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#16]
+	ldp	$tmpw2,$tmpw3,[$state,#24]
+	dup	@mx[4].s,$tmpw0
+	dup	@mx[5].s,$tmpw1
+	dup	@mx[6].s,$tmpw2
+	dup	@mx[7].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#32]
+	ldp	$tmpw2,$tmpw3,[$state,#40]
+	dup	@mx[8].s,$tmpw0
+	dup	@mx[9].s,$tmpw1
+	dup	@mx[10].s,$tmpw2
+	dup	@mx[11].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state, #48]
+	ldp	$tmpw2,$tmpw3,[$state,#56]
+	dup	@mx[13].s,$tmpw1
+	dup	@mx[14].s,$tmpw2
+	dup	@mx[15].s,$tmpw3
+___
+}
+
+sub sve_handle_blocks() {
+	my ($counter) = ("x10");
+
+	&SVE_LOAD_STATES();
+
+$code.=<<___;
+	cmp	$blocks,$veclen
+	b.ge	2f
+	// predicated processing
+___
+$code.=<<___;
+	mov	$counter,#10
+.align	5
+1:
+___
+	&SVE_INNER_BLOCK(0);
+$code.=<<___;
+	subs	$counter,$counter,1
+	b.ne	1b
+	b	1f
+2:
+	// unpredicated processing
+___
+$code.=<<___;
+	mov	$counter,#10
+.align	5
+2:
+___
+	&SVE_INNER_BLOCK(1);
+$code.=<<___;
+	subs	$counter,$counter,1
+	b.ne	2b
+1:
+___
```

Comment:
Is there actually any benefit to having 2 versions of the inner block (one predicated and one unpredicated)?

It makes sense that you would only use the predicated parts for combining with plain/ciphertext, but it seems like needless complication and codesize to optionally predicate the INNER_BLOCK logic.

Reply:
as demonstrated on latest micro-architecture, predicate instruction has significant overhead (10%+). It means if SVE register is saturated (or every element utilized), we are better off using unpredicated instruction. 

some instruction for inner-block (e.g. XAR etc.) will be unpredicated anyway, but other instructions like EOR, ADD, etc. need to avoid predication if it is not necessary.

Reply:
I'm saying why not _only_ have the unpredicated version for INNER_BLOCK?

Reply:
Agree that the decision should be made based on data.

I suspect that for V1 there is no real benefit to predication for 5/6/7 blocks - if you have 5 or more active .S lanes I would expect each SVE instruction needs to be split into 2 128b micro-operations (uops), and this would have the same impact on the backend as an unpredicated instruction (if anything probably be a bit slower).

There could be a benefit when:
(block_count % 8) is in the range [1,4] (i.e. 9...12, 17...24, etc.)
so in the final iteration when the predicated version is used, each SVE instruction only needs to produce 1 128b uop. It's not clear to me from the V1 SWOG that this actually happens.

Of course using predication in the tail case may show a bigger improvement for larger vector lengths depending on how the microarchitecture is set up (i.e. if predication can reduce uop count). In the long term I would expect that most of the time 1 SVE instruction will be 1 uop regardless of predication.

Reply:
let me run some test for the scenario where (block_count %8) in the range of [1, 4], it is probably easier to see the impact from test.

for foreseeable future, perhaps SVE vector length will mostly be 128/256 bits, (the only thing bigger that i am aware of, is a Fujitsu super computer on 512 bits). At the moment, we are only able to access 256-bit V1 hardware for testing.

Reply:
i think easier way is just to always handle small blocks or tail (that is, smaller than SVE lanes) in Neon. SVE will only handle full size. from performance point of view, perhaps this is the main scenario that matters, and could make code simpler.

Reply:
Yeah, I suppose that approach might work out better in practice, at least until SVE vector lengths are normally >256b - the ASIMD code handling tail cases needs to be in the OpenSSL binary anyway.
There is some elegance to having SVE predication handle the unexpected buffer lengths cleanly at the point of loading, xor-ing the buffer with the state, and storing, but it might not be enough faster to be worth the additional complexity.